File: patternlets/java-OpenMPI/22.scatterLoopGather/ScatterLoopGather.java
/* ScatterLoopGather.java
 * ... uses MPI to scatter a buffer of values into chunks, 
 *      a loop to process those chunks, and 
 *      a gather to combine the piecemeal values.
 *
 * Goal: The master process fills a buffer with values
 *        and scatters it to all the other processes.
 *       Each process doubles the values in its buffer-chunk.
 *       All processes then gather the chunks back to the master.
 *
 * Joel Adams, Calvin University, November 2019.
 *
 * Note: This assumes BUFFER_SIZE is evenly divisible by N.
 *
 * Usage: mpirun -np 4 java ./BroadcastLoopGather
 *
 * Exercise:
 * - Compile, then run, using 1, 2, 4, and 8 processes
 * - Use source code to trace execution and output
 * - Explain behavior/effect of the scatter and gather.
 * - Optional: change BUFFER_SIZE to be another multiple of 8, such as 16
 */

import java.nio.IntBuffer;
import mpi.*;

public class ScatterLoopGather {

 public static void main(String [] args) throws MPIException {
    MPI.Init(args);

    Comm comm               = MPI.COMM_WORLD;
    int id                  = comm.getRank();
    int numProcesses        = comm.getSize();
    IntBuffer scatterBuffer = null;
    IntBuffer chunkBuffer   = null;
    IntBuffer gatherBuffer  = null;

    if ( BUFFER_SIZE % numProcesses != 0 || numProcesses > BUFFER_SIZE ) {
        String errorMsg = "\nPlease run this program with -np N where N is\n"
                         + " <= " + BUFFER_SIZE + " and divides evenly into "
                         + BUFFER_SIZE + "\n\n";
        System.err.println(errorMsg);
        MPI.Finalize();
        System.exit(0);
    }

    if ( id == MASTER ) { 
        scatterBuffer = MPI.newIntBuffer(BUFFER_SIZE);
        fill(scatterBuffer);
        gatherBuffer = MPI.newIntBuffer(BUFFER_SIZE);
    }

    printBuffers("BEFORE the scatter", id, scatterBuffer, chunkBuffer, gatherBuffer);

    int chunkSize = BUFFER_SIZE / numProcesses;
    chunkBuffer = MPI.newIntBuffer(chunkSize);

    comm.scatter(scatterBuffer, chunkSize, MPI.INT, 
                  chunkBuffer, chunkSize, MPI.INT, MASTER);

    printSeparator("----", id);
    printBuffers("AFTER the scatter", id, scatterBuffer, chunkBuffer, gatherBuffer);

    doubleChunk(chunkBuffer);

    printSeparator("----", id);
    printBuffers("AFTER the doubling", id, scatterBuffer, chunkBuffer, gatherBuffer);

    comm.gather(chunkBuffer, chunkSize, MPI.INT, 
                 gatherBuffer, chunkSize, MPI.INT, MASTER);

    printSeparator("----", id);
    printBuffers("AFTER the gather:", id, scatterBuffer, chunkBuffer, gatherBuffer);

    MPI.Finalize();
  }

  /* utility to fill a Buffer with some values.
   * @param: buf, an IntBuffer.
   * POST: buf has been filled with int values.
   */
  private static void fill(IntBuffer buf) {
     for (int i = 0; i < buf.capacity(); ++i) {
         buf.put(i, i + 11);
     }
  }

  /* utility to print a buffer with labels.
   * @param: label, a String.
   * @param: id, an int containing the MPI rank of this process.
   * @param: sBuf, the IntBuffer the master fills and scatters.
   * @param: cBuf, the IntBuffer for storing a process's chunk.
   * @param: gBuf, the IntBuffer for storing the gathered results.
   * POST: The buffers' contents have been printed, with labels.
   */
  private static void printBuffers(String label, int id, 
                                    IntBuffer sBuf, IntBuffer cBuf, IntBuffer gBuf) {
    String msg = label + ", process " + id
                       + "'s scatterBuffer is: [";
    if (sBuf != null) {
        for (int i = 0; i < sBuf.capacity(); ++i) {
            msg += sBuf.get(i);
            if (i < sBuf.capacity()-1) msg += ",";
        }
    }
    msg += "]\n\t\t\t\tchunkBuffer is: [";
    if (cBuf != null) {
        for (int i = 0; i < cBuf.capacity(); ++i) {
            msg += cBuf.get(i);
            if (i < cBuf.capacity()-1) msg += ",";
        }
    }
    msg += "]\n\t\t\t\tgatherBuffer is: [";
    if (gBuf != null) {
        for (int i = 0; i < gBuf.capacity(); ++i) {
            msg += gBuf.get(i);
            if (i < gBuf.capacity()-1) msg += ",";
        }
    }
    msg += "]\n";
    System.out.print(msg);
  }

  /* utility to double the values in a chunk of an array.
   * @param: fullBuf, an IntBuffer containing all the values.
   * @param: id, the MPI rank of this process.
   * @param: chunkBuf, an IntBuffer into which we will write our values.
   * PRE: chunkBuf.capacity() == BUFFER_SIZE / numProcesses.
   * POST: chunkBuf contains the doubled values of this process's chunk
   *        of fullBuf.
   */
  private static void doubleChunk(IntBuffer chunkBuf) {
      for (int i = 0; i < chunkBuf.capacity(); ++i) {
          int value = chunkBuf.get(i);
          chunkBuf.put(i, value * 2);
      }
  }

  /* utility to print a separator string between the 'before' and 'after' parts.
   * @param: separator, a String.
   * @param: id, the rank of this MPI process.
   * POST: the master has printed the separator to System.out.
   */
  public static void printSeparator(String separator, int id) throws MPIException {
     MPI.COMM_WORLD.barrier();
     if (id == MASTER) { System.out.println(separator); }
     MPI.COMM_WORLD.barrier();
  }

  private static final int MASTER      = 0;
  private static final int BUFFER_SIZE = 8;
}

File: patternlets/MPI/23.scatterLoopGather/scatterLoopGather.c
/* scatterLoopGather.c
 * ... scatters an array of data into equal-sized chunks, 
 *      has each process use a loop to double the values in its chunk,
 *      and then gathers the chunks back to the master process.
 *
 * Joel Adams, Calvin University, December 2019.
 *
 * Precondition: ARRAY_SIZE is evenly divisible by N
 *               && N <= ARRAY_SIZE.
 *
 * Note: The output of different process's steps will be interleaved
 *       (even using barriers) b/c stdout is buffered
 *       and MPI does not guarantee FIFO output behavior.
 *
 * Usage: mpirun -np N ./scatterLoopGather
 *
 * Exercise:
 * - Compile and run, using 1, 2, 4, and 8 processes
 * - Use source code to trace execution and output
 * - Explain behavior/effect of MPI_Scatter(), MPI_Gather().
 * - Optional: change ARRAY_SIZE to be another multiple of 8, such as 16
 * - Optional: add calls to print() to display each array at each step
 */

#include <stdio.h>     // printf
#include <stdlib.h>    // malloc, exit, ...
#include <mpi.h>       // MPI functionality

#define MASTER     0
#define ARRAY_SIZE 8

void fill(int* a, int size);
void printSeparator(const char* separator, int id);
void print(char* locLabel, int id, char* aName, int* a, int numElements);

/*
 *  Main function: double the values in an array
 *  by dividing the work equally among N processes.
 */
int main(int argc, char** argv) {
    int* scatterArray = NULL;
    int* chunkArray = NULL;
    int* gatherArray = NULL;
    int numProcs = -1, myRank = -1, chunkSize = -1;

    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &numProcs);
    MPI_Comm_rank(MPI_COMM_WORLD, &myRank);

    printSeparator("", myRank);

    if (ARRAY_SIZE % numProcs || numProcs > ARRAY_SIZE) {
        if (myRank == MASTER) {
            printf("Please run with -np N divisible by and less than or equal to %d\n.", ARRAY_SIZE);
        }
        MPI_Finalize();
        exit(0);
    }

    if (myRank == MASTER) {     
        scatterArray = (int*) malloc( ARRAY_SIZE * sizeof(int) ); // allocate input array
        fill(scatterArray, ARRAY_SIZE);                           // populate it 
        gatherArray = (int*) malloc( ARRAY_SIZE * sizeof(int) );  // allocate result array
    }

    print("BEFORE Scatter", myRank, "scatterArray", scatterArray, ARRAY_SIZE);
    
    chunkSize = ARRAY_SIZE / numProcs;
    chunkArray = (int*) malloc(chunkSize * sizeof(int));          // allocate chunk array

    MPI_Scatter(scatterArray, chunkSize, MPI_INT,                 // scatter input array 
                 chunkArray, chunkSize, MPI_INT, 0, MPI_COMM_WORLD);

    print("AFTER Scatter", myRank, "chunkArray", chunkArray, chunkSize);

    for (unsigned i = 0; i < chunkSize; ++i) {                    // compute using chunk
        chunkArray[i] *= 2;
    }

    print("AFTER doubling", myRank, "chunkArray", chunkArray, chunkSize);

    MPI_Gather(chunkArray, chunkSize, MPI_INT,                   //  gather chunks
                gatherArray, chunkSize, MPI_INT,                 //   into gatherArray
                0, MPI_COMM_WORLD);

    print("AFTER gather", myRank, "gatherArray", gatherArray, ARRAY_SIZE);

    free(chunkArray);                                            // everyone clean up
    if (myRank == 0) {                                           // master clean up
        free(gatherArray); 
        free(scatterArray);
    }

    printSeparator("", myRank);

    MPI_Finalize();
    return 0;
}


/* fill an array with some easy-to-check values
 * @param: a, an int*.
 * @param: size, an int.
 * Precondition: a is the address of an array of ints.
 *              && size is the number of ints a can hold.
 * Postcondition: a has been filled with the values
 *                { 11, 12, 13, ... }.
 */
void fill(int* a, int size) {
    for (int i = 0; i < size; ++i) {
        a[i] = i+11;
    }
}

/* display a separator, synchronizing all processes
 * @param: separator, a char* 
 * @param: id, an int.
 * Precondition: separator points to a string to be be displayed
 *               && id is the MPI rank of this process.
 * Postcondition: separator has been displayed
 *                 and all MPI processes have been syncronized.
 */
void printSeparator(const char* separator, int id) {
    MPI_Barrier(MPI_COMM_WORLD);
    if (id == MASTER) { 
        printf("%s\n", separator);
    }
    MPI_Barrier(MPI_COMM_WORLD);
}

/* display a string, a process id, and its array values
 * @param: locLabel, a char*
 * @param: id, an int
 * @param: aName, a char*
 * @param: a, an int*.
 * @param: numElements, an int.
 * Precondition: locLabel points to a string describing our location
 *              && id is the rank of this MPI process
 *              && aName is the name of the array being printed
 *              && a is the address of an int array 
 *              && numElements is the number of int-values in a.
 * Postcondition: str, id, and a have all been written to stdout.
 */
void print(char* locLabel, int id, char* aName, int* a, int numElements) {
    printf("%s, process %d has this %s: {", locLabel, id, aName);
    if (a != NULL) {
        for (int i = 0; i < numElements - 1; ++i) {
            printf("%d, ", a[i]);
        }
        printf("%d}\n", a[numElements - 1]);
    } else {
        printf("}\n");
    }
}


